modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/PrepareServiceSelfTest.java

```diff
@@ -0,0 +1,294 @@
+import java.util.Map;
+import java.util.stream.Stream;
+import org.apache.calcite.runtime.CalciteContextException;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.ignite.internal.schema.NativeTypes;
+import org.apache.ignite.internal.sql.engine.QueryContext;
+import org.apache.ignite.internal.sql.engine.SqlQueryType;
+import org.apache.ignite.internal.sql.engine.framework.TestBuilders;
+import org.apache.ignite.internal.sql.engine.prepare.IgnitePlanner;
+import org.apache.ignite.internal.sql.engine.prepare.PlannerHelper;
+import org.apache.ignite.internal.sql.engine.prepare.PrepareServiceImpl;
+import org.apache.ignite.internal.sql.engine.prepare.QueryOptimizer;
+import org.apache.ignite.internal.sql.engine.prepare.ddl.DdlSqlToCommandConverter;
+import org.apache.ignite.internal.sql.engine.rel.IgniteRel;
+import org.apache.ignite.internal.sql.engine.sql.IgniteSqlParser;
+import org.apache.ignite.internal.sql.engine.sql.StatementParseResult;
+import org.apache.ignite.internal.sql.engine.trait.IgniteDistributions;
+import org.apache.ignite.internal.sql.engine.util.BaseQueryContext;
+import org.apache.ignite.sql.SqlException;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mockito;
+
+/**
+ * Prepare service self test.
+ */
+public class PrepareServiceSelfTest extends AbstractPlannerTest {
+    private final QueryContext queryCtx = QueryContext.create(SqlQueryType.ALL);
+    private PrepareServiceImpl service;
+
+    private QueryOptimizer queryPlannerSpy;
+    private DdlSqlToCommandConverter ddlPlannerSpy;
+
+    @BeforeEach
+    void setUp() {
+        queryPlannerSpy = Mockito.spy(new TestQueryOptimizer());
+        ddlPlannerSpy = Mockito.spy(new DdlSqlToCommandConverter(Map.of(), () -> "default"));
+
+        service = new PrepareServiceImpl(
+                "node",
+                10,
+                ddlPlannerSpy,
+                queryPlannerSpy
+        );
+
+        service.start();
+    }
+
+    @AfterEach
+    void tearDown() throws Exception {
+        service.stop();
+    }
+
+    @SuppressWarnings("WeakerAccess")
+    static Stream<Arguments> queries() {
+        return Stream.of(
+                // Query
+                Arguments.of("SELECT * FROM tbl WHERE id > 0", OBJECT_EMPTY_ARRAY),
+                // Query with args
+                Arguments.of("SELECT * FROM tbl WHERE id > ?", new Object[]{1}),
+                // DML
+                Arguments.of("INSERT INTO tbl VALUES (1, '42')", OBJECT_EMPTY_ARRAY),
+                Arguments.of("UPDATE tbl SET VAL = '42' WHERE id = 1", OBJECT_EMPTY_ARRAY),
+                // TODO IGNITE-19866: uncomment DELETE statement
+                // Arguments.of("DELETE FROM tbl WHERE id = 1", OBJECT_EMPTY_ARRAY),
```

Comment:
This issue will not reproduce if you add the PK column to the test table.

Reply:
I believe this should work without PK.